### PR TITLE
New: Make connectors provide charset & media type information

### PR DESCRIPTION
**Pull request by alrra**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [x] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/contributing/pull-requests/#commitmessages)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)

* Added ability for connectors to provide information about the response's charset and media type (this will be useful for future rules such as the compression and cache-control ones)
* Fixed issue regarding correctly determining the file extension.
* Made other minor improvements.

### CODE REVIEW

**Comment by molant** [comment]
> or at least, don't merge it until I finish the compression rule

Need this for the cache-control rule so hurry up 😛

**Comment by alrra** [comment]
> Need this for the cache-control rule so hurry up 😛
 
Well, this is is done, so I can be merged.

File: src/lib/connectors/utils/content-type.ts [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+import * as mimeDB from 'mime-db';
+import { parse, MediaType } from 'content-type';
+
+import { debug as d } from '../../utils/debug';
+import { IAsyncHTMLElement, IResponse } from '../../types';
+import { normalizeString } from '../../utils/misc';
+
+const debug = d(__filename);
+
+// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
```
Comment (molant) [inline src/lib/connectors/utils/content-type.ts]: Shouldn't this be a `Public/Private` banner?

File: src/lib/connectors/utils/content-type.ts [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    } catch (e) {
+        debug(`'content-type' header value is invalid (${e.message})`);
+
+        return null;
+    }
+
+    return contentType;
+};
+
+const getContentTypeData = (element: IAsyncHTMLElement, resource: string, response: IResponse) => {
```
Comment (molant) [inline src/lib/connectors/utils/content-type.ts]: Add `jsdoc` at least to `getContentTypeData` and `isTextMediaType` (the exported members)

File: src/lib/rules/content-type/content-type.ts [hunk 3]
```diff
@@ -246,14 +101,17 @@ const rule: IRuleBuilder = {
...
-                determineMediaTypeBasedOnElement(element) ||
-                determineMediaTypeBasedOnFileType(response.body.rawContent) ||
-                determineMediaTypeBasedOnFileExtension(resource);
+            /*
+             * Determined values
+             *
+             * Notes:
+             *
+             *  * The connectors already did all the heavy lifting here.
+             *  * For the charset, recomand `utf-8` for all text bases documents.
```
Comment (molant) [inline src/lib/rules/content-type/content-type.ts]: > recommend `utf-8` for all text based

**Pull request closed without merging**